p2p/gossip/protocol.go

```diff
@@ -0,0 +1,309 @@
+		morePeersReq: make(chan struct{}, config.RandomConnections),
+		peers:        make(map[string]*peer, config.RandomConnections),
+		oldMessageQ:  make(map[string]struct{}), // todo : remember to drain this
+		ps:           ps,
+		cp:           cp,
+	}
+}
+
+var _ Protocol = new(Neighborhood)
+
+type peer struct {
+	log.Log
+	node.Node
+	disc          chan error
+	connected     time.Time
+	conn          net.Connection
+	knownMessages map[string]struct{}
+	msgQ          chan []byte
+}
+
+func makePeer(node2 node.Node, c net.Connection, log log.Log) *peer {
+	return &peer{
+		log,
+		node2,
+		make(chan error,1),
+		time.Now(),
+		c,
+		make(map[string]struct{}),
+		make(chan []byte, PeerMessageQueueSize),
+	}
+}
+
+func (p *peer) send(message []byte) error {
+	if p.conn == nil || p.conn.Session() == nil {
+		return fmt.Errorf("the connection does not exist for this peer")
+	}
+	return p.conn.Send(message)
+}
+
+func (p *peer) addMessage(msg []byte) error {
+	// dont do anything if this peer know this msg
+	if _, ok := p.knownMessages[hex.EncodeToString(msg)]; ok {
+		return errors.New("already got this msg")
+	}
+
+	// check if connection and session are ok
+	c := p.conn
+	session := c.Session()
+	if c == nil || session == nil {
+		// todo: refresh Neighborhood or create session
+		return errors.New("no session")
+	}
+
+	data, err := message.PrepareMessage(session, msg)
+
+	if err != nil {
+		return err
+	}
+
+	select {
+	case p.msgQ <- data:
+
+	default:
+		return errors.New("Q was full")
+
+	}
+
+	return nil
+}
+
+func (p *peer) start(dischann chan struct{}) {
+	// check on new peers if they need something we have
+	//c := make(chan []string)
+	//t := time.NewTicker(time.Second * 5)
+	for {
+		select {
+		case m := <-p.msgQ:
+			err := p.send(m)
+			if err != nil {
+				// todo: handle errors
+				log.Error("Failed sending message to this peer %v", p.Node.PublicKey().String())
+				p.disc <- err
+			}
+		case d := <-p.disc:
+			log.Error("peer disconnected %v", d)
+			//p.conn.Close()
+			if dischann != nil {
+				dischann <- struct{}{}
+			}
+			return
+		}
+
+	}
+}
+
+func (s *Neighborhood) Shutdown() {
+	// no need to shutdown con, conpool will do so in a shutdown. the morepeerreq won't work
+	close(s.shutdown)
+}
+
+func (s *Neighborhood) Peer(pubkey string) (node.Node, net.Connection) {
+	s.peersMutex.RLock()
+	p, ok := s.peers[pubkey]
+	s.peersMutex.RUnlock()
+	if ok {
+		return p.Node, p.conn
+	}
+	return node.EmptyNode, nil
+
+}
+
+// the actual broadcast procedure, loop on peers and add the message to their queues
+func (s *Neighborhood) Broadcast(msg []byte) error {
+
+	if len(s.peers) == 0 {
+		panic("WHOHO")
+	}
+
+	s.oldMessageMu.RLock()
+	if _, ok := s.oldMessageQ[string(msg)]; ok {
+		// todo : - have some more metrics for termination
+		// todo	: - maybe tell the peer weg ot this message already?
+		return errors.New("old message")
+	}
+	s.oldMessageMu.RUnlock()
+
+	s.oldMessageMu.Lock()
+	s.oldMessageQ[string(msg)] = struct{}{}
```

Comment:
This isn't really a queue but more of a cache, though I called it a queue because we can't save old messages forever, I was planning this will be implemented as a fix sized "sliding-window". I have a todo to drain this. 

will open a issue for next PR.